File: systemProgramming/pythonAssembler/source/src/InstTable.py
import sys


class parsingError(Exception):
    def __str__(self):
        print("Parsing Error 발생")
        sys.exit()


class InstTable:
    instMap = {}

    def __init__(self):
        self.instMap = {}

    def __init__(self, instFile):
        self.instMap = {}
        self.openFile(instFile)

    def openFile(self, fileName):
        try:
            file = open(fileName, 'r')
            while True:
                line = file.readline()
                if not line: break
                inst = Instruction(line)
                self.instMap[inst.instruction] = inst
        except IOError as e:
            print(e)
            sys.exit()
        finally: file.close()

    def inst_info(self, instruction):
        if instruction[0] == '+':
            instruction = instruction[1:]
        if self.instMap.get(instruction) is None:
            return None
        return self.instMap[instruction]


class Instruction:

    def __init__(self, line):
        self.format = None
        self.instruction = None
        self.opcode = None
        self.n_operand = None
        self.parsing(line)

    def parsing(self, line):
        parts = line.split()
        for (i, part) in enumerate(parts):
            if i == 0:
                self.instruction = part
            elif i == 1:
                self.format = part
            elif i == 2:
                self.opcode = part
            elif i == 3:
                self.n_operand = int(part)
            else:
                raise parsingError

File: systemProgramming/javaAssembler_sim/source/src/SP20_simulator/InstTable.java
package SP20_simulator;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;


public class InstTable {

	HashMap<String, InstInfo> instMap;
	
	public InstTable() {
		instMap = new HashMap<String, InstInfo>();
	}
	
	public InstTable(String instFile) {
		instMap = new HashMap<String, InstInfo>();
		openFile(instFile);
	}
	

	public void openFile(String fileName) {
		// 파일 열어서 class 내부에 데이터 저장
		try {
			File file = new File(fileName);
			FileReader f_reader = new FileReader(file);
			BufferedReader b_reader = new BufferedReader(f_reader);
			String line = null;
			
			while((line = b_reader.readLine()) != null) {
				InstInfo inst = new InstInfo(line);
				instMap.put(inst.opcode, inst);
			}
			b_reader.close();
			} catch(FileNotFoundException e) {
				e.printStackTrace();
			} catch(IOException e) {
				e.printStackTrace();
			}
	}
	
	public InstInfo inst_info(String opcode) {
		if(!instMap.containsKey(opcode)) return null;
		InstInfo inst = instMap.get(opcode);
		return inst;
	}
	

}


class InstInfo {

	int n_operand;
	String format;
	String instruction;
	String opcode;
	
	public InstInfo(String line) {
		parsing(line);
	}
	
	public void parsing(String line) {
		//명령어 사전 line parsing
		String[] part = line.split("  ");
		
		for(int i = 0; i < part.length; i++) {
			if(i==0) 
				this.instruction = part[i];
			else if(i==1)
				this.format = part[i];
			else if(i==2)
				this.opcode = part[i];
			else
				this.n_operand = Integer.parseInt(part[i]);
		}
	}
}

